### ISSUE TITLE
Language detection interface

### ISSUE THREAD

**Issue by Cyclic3** [Ciphey/Ciphey#90]
### Problem
Whilst the original language checker is absolutely brilliant, it fails at small ciphertexts, or those with high entropy. An AI solution would be cool, but would be a bit OTT for rigid data structures such as json or CTF flags

### Solution
We present an interface for defining a custom candidate acceptor:
```py
class LanguageChecker(ABC):
    @staticmethod
    @abstractmethod
    def getArgs(**kwargs) -> dict: pass

    @abstractmethod
    def checkLanguage(self, text: str) -> bool: pass

ciphey_acceptor = LanguageCheckerDerived()
```

This can be passed as an argument to each cracker. Since a basic chi-squared filtration is applied by  the core to the candidates, relatively few should make it through. This allows the `is_acceptable` function to be a halt condition: if it returns true on any candidate, we stop.

However, a user shouldn't be expected to write such code for all text, that would be absurd. 

Instead, we provide a few language detection interfaces that can be selected by the user instead of them writing their own. My suggestion would be the following:

| Name | Args | Description |
| - | - | - |
| `brandon` | None  |  Current system, for long natural text |
| `neuralnet` | `path` | A neural net for text with deep structure, or smaller text |
| `regex` | `regex` | A regular expression for very simple structures |
| `user` | None | Prompts the user for each candidate |

I also put forward the following flags:
| Flag| Description |
| - | - |
| `-a <name>` | The internal language detector `<name>` (maybe with a configurable path) |
| `-A <path>` | A python script/module at `<path>` containing the `ciphey_acceptor` variable |
| `-p <param>=<value>` | Sets the kwarg `<param>` to `<value>` when the `is_acceptable` method is called |

**Comment by bee-san** [comment]
The `autumn` system will be rewritten to feature NLP, so DictionaryChecker is much faster due to lemmatisation. 

It would be nice to have an overall class where the user can state what level of accuracy they want, and for 1, 2, 3, 4 of these modules to run to ensure the extremely high accuracy the user wants.

For the `regex` system, I put forward that we should develop a Regex to match the format `flag{This is a flag}` which is ever so common in CTFs.

Another language detection interface would be a binary search over a list of cracked passwords like `rockyou.txt`. The text file will have to be pre-installed on the user's system, but we can use the default Kali Linux path `/usr/share/wordlists/rockyou.txt` to do so. Even though it is not anything smart, I still suspect it will come in very useful to CTF players.

I feel like passwords in rockyou, and the CTF flag system matches the two most likely use cases by CTF players.

The `regex` module should use some kind of settings file that the user can set. Perhaps a plain text file with values separated by `\n` to allow the user to create their own regexes. 

The `user` module should also use some kind of storage system.

**TL;DR:**
I am suggesting 1 new language interace:
* Binary search through  Rockyou.txt
A pre-built module for regex:
* CTF flag format `flag{hello}`
A new argument:
* -l <level> : How many levels of confirmation do you need?
The above argument will let the user decide if they want to run the text through X language detection interfaces to be 100% sure of the text they have being language.
* Regex module should allow users to "import" from a previous file.
* User module should use some kind of storage system.

Edit:
Another Regex that would be nice (or maybe not regex) would be pattern matching IP addresses / websites.

**Comment by Cyclic3** [comment]
Great suggestions, but I have a few comments:
1. Flag formats are so diverse (from hex strings in the braces to english sentances) that I still think a user supplied regex is best
2. Multiple langauge detction modules make the command line syntax really compliacatied (how do I differentiate between parameters for each module?) and can trivially be implemented in a python script
3. RockYou idea is great! However, we would have to download the 100MB file onto the user's computer, so maybe just have an aribtrary sorted dictioanry module?
4. What do you mean by storage system?

**Comment by brandonskerritt** [comment]
1. Pareto principle? 3 most popular CTf websites are HTB/THM/Vuln hub. We just do Regexs for those 3 and we hit 80% of all CTF flags? The user can add more if they want. We will only need to search for HTB{*}, THM{*}, VULN{*}.
2. I agree this one kinda sucks, more of a "would be nice but not super needed"
3. We don't have to download rockyou, if we assume the user is using Kali / Parrot / Seclists they have the text file at /usr/share/wordlists/rockyou.txt . If it's not there, we tell the user to download the file or supply a file path. Although I just realised Rockyou isn't sorted alphabetically so we'd have to store a local version and sort it and use that, as opening files is slow too.
4. If the user writes multiple regexes / candidates it would be nice for us to give them the option of storing it in like a text file so they can use it next time without worry of re-writing them.

**Comment by Cyclic3** [comment]
1\. Yeah, that sounds good. I'll put it in cipheydists so that it is more maintainable
3\. I really don't like seeking paths that only exist on a few OSs. I believe this is more suited for #94 
4\. By user prompting, I meant pass each candidate to the user, using the best trained neural net I know of (the human brain). It is more annoying, but the best way of making sure nothing slips through for high-entropy plaintexts

**Comment by bee-san** [comment]
For the **autumn** interface, here is a rough design of the new interface from #60 (now closed, as this issue is pinned).
* [ ] - Use Lemmisation to reduce dictionary checker size
* [ ] - Use Lemminisation in the text to check to increase accuracy
* [ ] - For check 1000 words, it may be better to search for stop words or word endings.
* [ ] - write tests for check1000words

**Comment by brandonskerritt** [comment]
@Cyclic3 Can you list the arguments in their fullnames, instead of letters? For example `-w = --wordlist`. This is for the settings file :smile:

**Comment by Cyclic3** [comment]
> @Cyclic3 Can you list the arguments in their fullnames, instead of letters? For example `-w = --wordlist`. This is for the settings file smile

@autumnskerritt Where do you want me to list them?

**Comment by brandonskerritt** [comment]
@Cyclic3 In the settings file settings.yml on your iface branch :)

**Comment by bee-san** [comment]
@Cyclic3 this is how `autumn` interface works. Any comments?

**Phase 1. Try to turn the text into something we can search with in phase 2**
Firstly, we turn the string into words, we perform lemmatization (greatly -> great) and then we turn it into a base form synonym (still working on this, but basically the idea is (great -> good) but not too sure if this is possible).

If we can do this, we pass phase 1. It is semi-likely that the text is English if we can perform these operations on it. Phase 1 isn't about accuracy but more about speed.

In total this is O(n) time, as we only need to loop over the words once to perform all actions.

**Phase 2.** Dictionary checker version 2. 
We have a new dictionary with reduced words (greatly -> great -> good). This makes the space complexity of the dictionary much smaller.

We then store the dictionary in the LC object as a dict, for O(1) lookup time. Since the space complexity is reduced, and we pas the LC object instead of create it everytime - this won't have a large effect on the program.

Since we have already performed the work required to compare our text to the dictionary in Phase 1, Phase 2 is effectively O(n) + O(1).

**Chunking in Phase 2**
However, I have split the text up into chunks. Since accuracy is improved (due to a larger dict with more unique words, synonym reduction and lemmatization) the threshold percentage is now much lower (25% - 35%, I need to optimise this). 

The text is split up into chunks of 25%. So in previous versions, we calculated 100% of the text and checked to see if 55% matched. Now, we only check 25%. If it matches, we return True. If not, we go to the next part and check to see if we go over 25%. 

Either we return early (most likely), we don't return at all (likely), or we return at the end of the string. 

On much longer texts (10,000+ words) this increases speed greatly compared to the older version. 

**Big O notation**

This means the entire LC checker is O(n) + O(n) = O(n^2) for the worst case possible.

**Final remarks**
I'm not 100% sure on the synonym thing as there is no pre-selected synonym for words. For example, "great -> good" but "brilliant -> great" but "good -> fantastic" kind of thing. 

Other than that, what do you think?

**Note: this exact model would work really well on the 59 languages that Spacy supports. We just need the dictionary files and we can add support for more languages easily, assuming crackers / decoders support the different chars**

**some late night edits**

the synonyms appear to be ordered, possibly by the frequency of the word itself which is a feature from Google Books. if the exact synonyms are non deterministic and going by order, I would assume brilliant will always equal good. this makes my job easier.

but I cannot guarantee this for every possible text.

If I have gotten this wrong and it is in fact non deterministic and they are not perfect mappings of eachother, we could generate a list of synonyms per word and use the dictionary lookup to see if that word appears. 

if we have to do that, I would argue we wouldn't need synonym checking at all and to remove it. I will have to explore further to see which is the case and whether or not it is worth it.

**Comment by Cyclic3** [comment]
That sounds absolutely brilliant! Just a few things I see:

1. You say O(n) + O(n) = O(n^2), but Im pretty sure that's O(n)

2. Does the lemmatisation use a dictionary or something, or is it just spooky AI magic? I ask because it sounds like we might be getting more information than data at that point, and that's never good.

3. Be aware that the stuff passed to the checker will most likely be complete gibberish (with a similar freq dist) OR the correct result. A user will not care about an extra second spent on the final correct result, but really will care that every false candidate takes an extra second. The current suggestion seems to be pessimal for the gibberish inputs: maybe add some sanity checks (have I failed to match any word, have I failed to lemmatise any word, etc.)

That being said, this looks like a significant improvement on the current autumn checker, and I look forward to it's implementation!

**Comment by brandonskerritt** [comment]
For point 3 before I forget, I am thinking of something like this:

accept = [ ] 
lemmitisztion (append 1 to accept)
word endings (simple if "ing, en," on end of a word) (append to list)
if synonyms (append to list)

phase 2 enters if list >= Len 2 

so we have lots of little checks, but together they should give a relatively clear picture on whether it's worth it to go to phase 2 or not. The exact checks are unknown, I just needed to write this down before I slept.

Also, I will make it so as soon as the list hits 2 or X amount, it will go to phase 2. I don't want to wait around when we've already accepted that it's going to phase 2. probably will use a While loop for this. like
funcs = [self.lem, word.ends, word.sym] 
I = 0
while accept <= 2:
    accept.append(1) if funcs[I] else Continue

this way we can easily add more smaller checks if we wanted to change something while maintaining the decorum of the program. 

will edit this comment tomorrow with full reply and DM you about it.

**Comment by Cyclic3** [comment]
I would suggest a simple lower bound test: we pass if we get more than 25%,and fail if we get lower than 5% (or smth idk) for n consecutive windows. 

You _can_ use the chi squared in cipheycore (which supports processing of split strings), but that would often be redundant. 

In the core, the vigenere and caesar crackers actually discard the string, only keeping the frequency analysis, and use the chi squared metric to filter the 26^n possible keys. Passing this unfiltered to python, and then letting python process it, would take truely huge amounts of time. This means that a chi squared metric *has* to be performed in the core. 

That being said, there can exist languages whose plaintexts fail chi squared analysis, like one where valid strings are only made up of a single character being repeated n times, where that character is chosen from a set of length k (for example: aaaaaaaa, eee, iiiiiii, o). A chi squared analysis would completely fail on this, giving a score of O(n^k), which fails for any p value given a long enough string.

Perhaps we can return an object from the cracker which states what tests have been performed, to save time on redundant analysis. With such information, autumn could make an intelligent decision to just use a wordlist if enough analysis was performed, and the more detailed analysis if it wasn't.

**Comment by bee-san** [comment]
If I'm reading this correctly:
> I would suggest a simple lower bound test: we pass if we get more than 25%,and fail if we get lower than 5% (or smth idk) for n consecutive windows.
You're suggesting that we run all tests and see if we get 25% Imo that would be much slower. What do you mean by `n windows`?

Okay, Chi squared is out then! 

> Perhaps we can return an object from the cracker which states what tests have been performed, to save time on redundant analysis. With such information, autumn could make an intelligent decision to just use a wordlist if enough analysis was performed, and the more detailed analysis if it wasn't.
This is entirely possible. I will add support to `autumn` checker to skip phase 1 if it receives an dictionary with key `"phase1": True` for `True == skip phase 1`. 

If you have more tests, let me know and I can factor them in.

In your first reply:
https://github.com/Ciphey/Ciphey/issues/90#issuecomment-645046918
Point 3:
> Be aware that the stuff passed to the checker will most likely be complete gibberish (with a similar freq dist) OR the correct result. A user will not care about an extra second spent on the final correct result, but really will care that every false candidate takes an extra second. The current suggestion seems to be pessimal for the gibberish inputs: maybe add some sanity checks (have I failed to match any word, have I failed to lemmatise any word, etc.)

I decided to test how well `lem` worked as phase 1. To do this, I created this program:
```python
"""
TL;DR

Tested over 20,000 times

Maximum sentence size is 15 sentences
1/2 chance of getting 'gibberish' (encrypted text)
1/2 chance of getting English text

Each test is timed using Time module.
The accuracy is calculated as to how many true positives we get over the entire run

"""


import spacy
import random
import time
from statistics import mean
import ciphey
import enciphey
from alive_progress import alive_bar

nlp = spacy.load("en_core_web_sm")

f = open("hansard.txt", encoding="ISO-8859-1").read()
f = f.split(".")

enciph = enciphey.encipher()


def lem(text):
    sentences = nlp(text)
    return set([word.lemma_ for word in sentences])


def get_random_sentence():
    if random.randint(0, 1) == 0:
        x = None
        while x is None:
            x = (True, " ".join(random.sample(f, k=random.randint(1, 50))))
        return x
    else:
        x = None
        while x is None:
            x = enciph.getRandomEncryptedSentence()
            x = x["Encrypted Texts"]["EncryptedText"]
        return (False, x)


# Now to time it and take measurements


def perform():
    # calculate accuracy
    total = 0
    true_returns = 0

    # calculate aveager time
    time_list = []

    # average sentance size
    sent_size_list = []
    items = range(20000)
    with alive_bar(len(items)) as bar:
        for i in range(0, 20000):
            sent = get_random_sentence()
            text = sent[1]
            truthy = sent[0]
            sent_size_list.append(len(text))

            # should be length of chars
            old = len(text)

            # timing the function
            tic = time.perf_counter()
            new = lem(text)
            tok = time.perf_counter()

            # checking for accuracy
            new = len(new)
            # the and here means we only count True Positives
            if new < old and truthy:
                true_returns += 1
            total += 1

            # appending the time
            t = tok - tic
            time_list.append(t)
            bar()

    print(
        f"The accuracy is {str((true_returns / total) * 100)} \n and the time it took is {str(round(mean(time_list), 2))}. \n The average string size was {str(mean(sent_size_list))}"
    )


perform()
```

The results were fascinating, to say the least.
With a 50/50 chance of the text being gibberish (ciphertext from enCiphey) or sentences from Hansard.txt, we had these results for using lemmization as phase 1:

```
The accuracy is 49.63%
 and the time it took is 0.02 seconds on average.
 The average string size was 1133.63255.
```

**We get a 50% accuracy with a speed of 0.02 seconds on average, across 20k tests with the average size of a string being 1133 chars. **

The accuracy is quite bad considering that a coin flip is 50/50.
On average, the user would expect Phase 2 to be entered 50% of the time, which is annoying as phase 2 is quite slow. But by itself it's quite fast.

I am going to build the "2nd phase" of phase 1 using the While Loop we saw earlier. If we can combine just one more metric, we would see much higher accuracy and again - likely incredibly low latency.

I will create a table of my results:

## Table of max sentence length == 50

| Name                       | Speed                        | Accuracy | String Size Average Chars | Epochs | Max Sentence Size |
| -------------------------- | ---------------------------- | -------- | ------------------------- | ------ | ----------------- |
| Lemmization (lem)          | 0.02 seconds                 | 50%      | 1580                      | 20,000 | 50                |
| Stop word removal          | 3.05465052884756e-05 seconds | 96%      | 1596                      | 20,000 | 50                |
| Check1000Words             | 0.0005 seconds               | 96%      | 1597                      | 20,000 | 50                |
| Word endings               | 0.0009 seconds               | 95%      | 1597                      | 20,000 | 50                |

## Table of max sentence length == 5

| Name                       | Speed                          | Accuracy | String Size Average Chars | Epochs | Max Sentence Size |
| -------------------------- | ------------------------------ | -------- | ------------------------- | ------ | ----------------- |
| Lemmization (lem)          |
| Stop word removal          | 1.1574924453998391e-05 seconds | 93%      | 569                       | 20,000 | 5                 |
| Check1000Words             | 0.0006 seconds                 | 95%      | 586                       | 20,000 | 5                 |
| Word endings               | 0.0003 seconds                 | 92%      | 482                       | 20,000 | 5                 |
## Table of max sentence length == 1

| Name                       | Speed                           | Accuracy | Threshold | String Size Average Chars | Epochs | Max Sentence Size |
| -------------------------- | ------------------------------- | -------- | ------ |------------------------- | ------ | ----------------- |
| Lemmization (lem)          |
| Stop word removal          | 1.2532061150591289e-05. seconds | 50%      | 481                       | 20,000 | 1                 |
| Check1000Words             | 0.0006 seconds                  | 95%      | 586                       | 20,000 | 5                 |
| Word endings               | 0.0002 seconds                  | 86%      | 15| 482                       | 20,000 | 1                 |


## Confusion Matrices & Notes
### Lemization

```
                Positive    Negative
Positive     10031      9967
Negative     2            0
```


### Stop Words
This test was performed where the text was not `.lower()`, so the actual accuracy _may_ be a little tiny bit higher since the stop words list is all lowercase.

50 sentence limit

```
                        Positive    Negative
            Positive     9913            855
            Negative     56            9176
```

5 sentence limit:

```
                        Positive    Negative
            Positive     9513            967
            Negative     530            8990
```

### Check 1000 words

50 sentence limit

```
                           Positive    Negative
            Positive     10008            552
            Negative     56            9384
```

5 sentence limit

```
                        Positive    Negative
            Positive     9563            597
            Negative     397            9443
```

# Analysis
**I believe that the best Autumn checker will look at the length of the text, and adjust the % threshold and the exact phase 1 checker per text.**

The below data is taken from calculations performed over many hours. it shows the best threshold % for the best phase 1 checker with the highest accuracy. These checkers were chosen as others showed a maximum accuracy of 58%.

```
{'check 1000 words': {1: {'Accuracy': 0.925, 'Threshold': 2},
                      2: {'Accuracy': 0.95, 'Threshold': 68},
                      3: {'Accuracy': 0.975, 'Threshold': 62},
                      4: {'Accuracy': 0.98, 'Threshold': 5},
                      5: {'Accuracy': 0.985, 'Threshold': 54}},
 'stop words': {1: {'Accuracy': 0.865, 'Threshold': 50},
                2: {'Accuracy': 0.93, 'Threshold': 19},
                3: {'Accuracy': 0.965, 'Threshold': 15},
                4: {'Accuracy': 0.97, 'Threshold': 28},
                5: {'Accuracy': 0.985, 'Threshold': 29}}
```

Where the numbers are:

```
1 : The mean is 87.62
2 : The mean is 110.47925
3 : The mean is 132.20016666666666
4 : The mean is 154.817125
5 : The mean is 178.7297
```

Looking at this test, it is clear that stopwords is better than check 1000 words for speed, but the accuracy is a little bit slower. Stop words is incredibly faster than check 1k words, but on a smaller input the stopwords checker breaks.

Therefore, we should use stopword checker on larger texts, and check 1k words on smaller texts.

More specifically, stopwords checker for len == 110 has an optimal threshold of 19, whereas check 1k words has an optimal threshold of 68. This means that while stopwords can potentially end earlier and only search the first 19% of the list, check 1k words would search 68% of the list.

Stopwords has a lower accuracy by 2%, but it is much, much faster and its optimal threshold is greatly reduced. 

We would have this algorithm:
1. Sentence length less than 110:
   1. Use check 1k words with threshold of 2%
2. Sentence length > 110:
   1. use Stopwords with threshold of 15
3. Sentence length > 150:
   1. Stopwords threshold increases to 28
   
This is the ideal optimal phase 1 algorithm for `autumn` checker.


# TODO
* [ ] Use the testing system in place to find the optimal percentage thresholds for each phase of language checker. I can use the confusion matrices to help work this out.
* [ ] Only perform phase 1 on {THRESHOLD}% of list etc
* [ ] Look at using different threshold numbers for different lengths of text, and wether or not that dramatically increases accuracy.
* [ ] Calculate speed / accuracy of old language Checker
* [ ] Calculate speed / accuracy of Harlan's neural network
* [ ] Clean this issue into documentation after I'm done

**Comment by Cyclic3** [comment]
Closed for #104
